### PR TITLE
New `join_cluster` extrinsic

### PR DESCRIPTION
**Pull request by khssnv**
## Description

New `join_cluster` extrinsic for the `pallet-ddc-clusters` which allows node provider to add a node to a cluster if it can pass `node_auth_smart_contract` authorization. It completes an unattended cluster extension flow implementation for the cluster manager allowing them to express cluster extension rules in form of smart contract requiring no further transactions from their account to add nodes to the cluster.

## Types of Changes
Please select the branch type you are merging and fill in the relevant template.

- [ ] Hotfix
- [ ] Release
- [x] Fix or Feature

### Types of Changes

- [ ] Tech Debt (Code improvements)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Dependency upgrade (A change in substrate or any 3rd party crate version)

### Checklist for Fix or Feature

- [x] Change has been tested locally.
- [x] Change adds / updates tests if applicable.
- [x] Changelog doc updated.
- [x] `spec_version` has been incremented.
- [ ] `network-relayer`'s [events](https://github.com/Cerebellum-Network/network-relayer/blob/dev-cere/shared/substrate/events.go) have been updated according to the blockchain events if applicable.
- [x] All CI checks have been passed successfully

### CODE REVIEW

File: pallets/ddc-clusters/src/lib.rs [hunk 1]
```diff
@@ -394,6 +395,56 @@ pub mod pallet {
 
 			Self::do_validate_node(cluster_id, node_pub_key, succeeded)
 		}
+
+		#[pallet::call_index(5)]
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::add_node())]
+		// #[pallet::weight(<T as pallet::Config>::WeightInfo::join_cluster())]
```
[the remark this answers is not in the document]
Comment (khssnv) [reply pallets/ddc-clusters/src/lib.rs]: Thanks for noticing, updated.

File: pallets/ddc-clusters/src/tests.rs [hunk 2]
```diff
@@ -126,7 +126,7 @@ fn create_cluster_works() {
 }
 
 #[test]
-fn add_and_delete_node_works() {
+fn add_join_and_delete_node_works() {
```
[the remark this answers is not in the document]
Comment (khssnv) [reply pallets/ddc-clusters/src/tests.rs:129]: Updated, all significant cases covered.

File: pallets/ddc-clusters/src/lib.rs [hunk 3]
```diff
@@ -538,6 +588,44 @@ pub mod pallet {
...
+
+			node.set_cluster_id(Some(cluster.cluster_id));
+			T::NodeRepository::update(node).map_err(|_| Error::<T>::AttemptToAddNonExistentNode)?;
+
+			ClustersNodes::<T>::insert(
+				cluster.cluster_id,
+				node_pub_key.clone(),
+				ClusterNodeState {
+					kind: ClusterNodeKind::External,
+					status: ClusterNodeStatus::ValidationSucceeded,
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/lib.rs:609]: Do we want to treat a node that is being added to the cluster as operational/trusted from the very beginning (i.e. set `ValidationSucceeded` status immediately)? 

Ideally, it should have `AwaitsValidation` status right after it is added to the cluster, and then validated by another nodes within a cluster by some challenge. After passing (or not passing) the challenge, the status should be automatically (by validators OCWs/DAC) changed to `ValidationSucceeded` (or `ValidationFailed`).

Currently, we are missing this "operability" challenge on both DDC and blockchain sides, and that's why the validation is performed manually by sending the `validate_node` extrinsic on behalf Cluster Manager account.

Won't it be an issue if `add_node` extrinsic and `join_cluster` extrinsic lead to different node status within a cluster ? As far as I know, Cluster Manager UI utilizes the `add_node`  extrinsic for adding node to cluster on behalf cluster manager account

File: pallets/ddc-clusters/src/lib.rs [hunk 4]
```diff
@@ -394,6 +395,55 @@ pub mod pallet {
 
 			Self::do_validate_node(cluster_id, node_pub_key, succeeded)
 		}
+
+		#[pallet::call_index(5)]
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::join_cluster())]
+		pub fn join_cluster(
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/lib.rs:401]: As we introduced the `join_cluster` extrinsic, let's remove checks and calls to the `node_provider_auth_contract_address` within the `add_node` extrinsic body to eliminate ambiguities of the `node_provider_auth_contract` main purpose.

The main goal of this contract is to check whether a node provided is allowed to join a cluster without waiting for action from the cluster manager. In the scope of the `add_node` extrinsic, the contract validation doesn't make sense as the Cluster Manager is permitted to change the contract (address) whenever it wants to. So the checks in the contract should be performed only in the scope of `join_cluster` extrinsic

File: pallets/ddc-clusters/src/weights.rs [hunk 5]
```diff
@@ -88,6 +89,45 @@ impl<T: frame_system::Config> WeightInfo for SubstrateWeight<T> {
...
+	// Proof: `System::EventTopics` (`max_values`: None, `max_size`: None, mode: `Measured`)
+	// Storage: `DdcClusters::ClustersNodesStats` (r:1 w:1)
+	// Proof: `DdcClusters::ClustersNodesStats` (`max_values`: None, `max_size`: None, mode: `Measured`)
+	// Storage: `DdcClusters::ClustersNodes` (r:0 w:1)
+	// Proof: `DdcClusters::ClustersNodes` (`max_values`: None, `max_size`: None, mode: `Measured`)
+	// Storage: UNKNOWN KEY `0x89eb0d6a8a691dae2cd15ed0369931ce0a949ecafa5c3f93f8121833646e15c3` (r:1 w:0)
+	// Proof: UNKNOWN KEY `0x89eb0d6a8a691dae2cd15ed0369931ce0a949ecafa5c3f93f8121833646e15c3` (r:1 w:0)
+	// Storage: UNKNOWN KEY `0xc3ad1d87683b6ac25f2e809346840d7a7ed0c05653ee606dba68aba3bdb5d957` (r:1 w:0)
+	// Proof: UNKNOWN KEY `0xc3ad1d87683b6ac25f2e809346840d7a7ed0c05653ee606dba68aba3bdb5d957` (r:1 w:0)
+	fn join_cluster() -> Weight {
```
Comment (yahortsaryk) [inline pallets/ddc-clusters/src/weights.rs:126]: Did you use the benchmarking machine provided by devops to generate the weights ?

**Review by yahortsaryk — CHANGES_REQUESTED** [review]

File: pallets/ddc-clusters/src/lib.rs (hunk 3 above)
Comment (khssnv) [reply pallets/ddc-clusters/src/lib.rs:609]: > Do we want to treat a node that is being added to the cluster as operational/trusted from the very beginning (i.e. set ValidationSucceeded status immediately)?

Yes, we want it operational from the very beginning at least for now while we have no other means to check it as you mentioned. It also passed the check by the smart contract which may contain an off-chain flow with the custom connection check implemented by the cluster manager.

> Won't it be an issue if add_node extrinsic and join_cluster extrinsic lead to different node status within a cluster ?

I don't see any issue right now. Please let me know if you see any.

File: pallets/ddc-clusters/src/lib.rs (hunk 4 above)
Comment (khssnv) [reply pallets/ddc-clusters/src/lib.rs:387]: The `node_provider_auth_contract` call removed from `add_node`.

**Review by yahortsaryk — APPROVED** [review]

**Pull request merged**